src/common/substrait/src/df_expr.rs

```diff
@@ -0,0 +1,714 @@
+            let high = expression_from_df_expr(ctx, high, schema)?;
+            let arguments = utils::expression_to_argument(vec![expr, low, high]);
+            let op_name = if *negated { "not_between" } else { "between" };
+            let function_reference = ctx.register_scalar_fn(op_name);
+            utils::build_scalar_function_expression(function_reference, arguments)
+        }
+        // Don't merge them with other unsupported expr arms to perserve the ordering.
+        Expr::Case { .. } | Expr::Cast { .. } | Expr::TryCast { .. } => UnsupportedExprSnafu {
+            name: expr.to_string(),
+        }
+        .fail()?,
+        Expr::Sort {
+            expr,
+            asc,
+            nulls_first: _,
+        } => {
+            let expr = expression_from_df_expr(ctx, expr, schema)?;
+            let arguments = utils::expression_to_argument(vec![expr]);
+            let op_name = if *asc { "sort_asc" } else { "sort_des" };
+            let function_reference = ctx.register_scalar_fn(op_name);
+            utils::build_scalar_function_expression(function_reference, arguments)
+        }
+        Expr::ScalarFunction { fun, args } => {
+            let arguments = utils::expression_to_argument(
+                args.iter()
+                    .map(|e| expression_from_df_expr(ctx, e, schema))
+                    .collect::<Result<Vec<_>>>()?,
+            );
+            let op_name = utils::name_builtin_scalar_function(fun);
+            let function_reference = ctx.register_scalar_fn(op_name);
+            utils::build_scalar_function_expression(function_reference, arguments)
+        }
+        // Don't merge them with other unsupported expr arms to perserve the ordering.
+        Expr::ScalarUDF { .. }
+        | Expr::AggregateFunction { .. }
+        | Expr::WindowFunction { .. }
+        | Expr::AggregateUDF { .. }
+        | Expr::InList { .. }
+        | Expr::Wildcard => UnsupportedExprSnafu {
+            name: expr.to_string(),
+        }
+        .fail()?,
+    };
+
+    Ok(expression)
+}
+
+/// Convert DataFusion's `Column` expr into substrait's `FieldReference` -
+/// `DirectReference` - `StructField`.
+pub fn convert_column(column: &Column, schema: &Schema) -> Result<FieldReference> {
+    let column_name = &column.name;
+    let field_index = schema
+        .column_index_by_name(column_name)
+        .context(MissingFieldSnafu {
+            field: format!("{:?}", column),
+            plan: format!("schema: {:?}", schema),
+        })?;
+
+    Ok(FieldReference {
+        reference_type: Some(FieldReferenceType::DirectReference(ReferenceSegment {
+            reference_type: Some(SegReferenceType::StructField(Box::new(StructField {
+                field: field_index as _,
+                child: None,
+            }))),
+        })),
+        root_type: None,
+    })
+}
+
+/// Some utils special for this `DataFusion::Expr` and `Substrait::Expression` convertion.
+mod utils {
```

Comment:
`name_builtin_scalar_function` and `name_df_operator` should be added to DataFusion. I'll do this later.